keps/sig-scheduling/5710-workload-aware-preemption/README.md

```diff
@@ -0,0 +1,1064 @@
+e.g. preempt our workload. But intuition is not enough here.
+
+As described in user stories above, a simple static priority doesn't seem to be enough. Arguably it is
+not even a single priority because a priority used for scheduling can be different than the priority
+that should be used for preemption. So in the ideal world a workload owner should be able to define:
+
+- priority used for scheduling (potentially also separately for every PodGroup)
+- priority used for preemption (again potentially also for every PodGroup) and be able to arbitrarily
+  mutate it during the whole lifecycle of the workload
+
+We start simpler though and assume that all PodGroups have the same scheduling and preemption
+policy by extending the `Workload` API as following:
+
+```golang
+type WorkloadSpec struct {
+    // Existing field(s).
+
+    // PriorityClassName, if specified, indicates the workload's priority that
+    // should be used when scheduling this workload. "system-node-critical" and
+	// "system-cluster-critical" are two special keywords which indicate the
+	// highest priorities with the former being the highest priority. Any other
+	// name must be defined by creating a PriorityClass object with that name.
+	// If not specified, the priority will be default or zero if there is no
+	// default.
+    //
+    // This field is immutable.
+    PriorityClassName *string
+
+    // PreemptionPriorityClassName, if specified, indicates the workload's
+    // priority that should be used when attempting to preempt this workload.
+    // If not specified, it will default to PriorityClassName.
+    //
+    // This field is mutable.
+    PreemptionPriorityClassName *string
+}
+```
+
+If that appears not being enough, we will similarly extend the `PodGroup` API in a follow up.
+In such case, the priority of a pod would be the priority of a smallest unit in the `Workload`
+object (Workload, PodGroup, PodSubGroup, ...) corresponding to this pod.
+
+
+There is one direct implication of the above - the `pod.Spec.PriorityClassName` and `pod.Spec.Priority`
+may no longer reflect the actual pod priority. This can be misleading to users.
+
+```
+<<[UNRESOLVED priority divergence]>>
+There are several options we can approach it (from least to most invasive):
+- Explain via documentation
+- Validating that if a pod is referencing a workload, `pod.Spec.PriorityClassName` equals
+  `workload.Spec.PriorityClassName`. However, `Workload` object potentially may not exist
+  yet on pod creation.
+- Making `pod.Spec.PriorityClassName` and `pod.Spec.Priority` mutable fields and having a
+  controller responsible for reconciling these. However, that doesn't fully address the
+  problems as divergence between the pod and PodTemplate in true workload object could also
+  be misleading.
+
+The validation option seems like the best option, if we can address the problem of not-yet
+existing `Workload` object (reversed validation?).
+<<[/UNRESOLVED]>>
+```
+
```

Comment:
Should PriorityClass.globalDefault affect Workload priority the same way it affects pod priority?

Should PriorityClass.preemptionPolicy apply to a priority class mentioned in a Workload?